core/emulate.c

```diff
@@ -920,17 +987,63 @@ static void em_movsx(struct em_context_t *ctxt)
     ctxt->dst.value = 0;
     memcpy(&ctxt->dst.value, &extension, ctxt->dst.size);
     memcpy(&ctxt->dst.value, &ctxt->src1.value, ctxt->src1.size);
+    return EM_CONTINUE;
+}
+
+static em_status_t em_push(struct em_context_t *ctxt)
+{
+    em_status_t rc;
+    em_operand_t sp;
+
+    // Pre-decrement stack pointer
+    register_add(ctxt, REG_RSP, -(int64_t)ctxt->operand_size);
+
+    memset(&sp, 0, sizeof(sp));
+    sp.type = OP_MEM;
+    sp.size = ctxt->operand_size;
+    sp.mem.ea = gpr_read(ctxt, REG_RSP, ctxt->address_size);
+    sp.mem.seg = SEG_SS;
+    sp.value = ctxt->src1.value;
+
+    rc = operand_write(ctxt, &sp);
+    return rc;
+}
+
+static em_status_t em_pop(struct em_context_t *ctxt)
+{
+    em_status_t rc;
+    em_operand_t sp;
+
+    memset(&sp, 0, sizeof(sp));
+    sp.type = OP_MEM;
+    sp.size = ctxt->operand_size;
+    sp.mem.ea = gpr_read(ctxt, REG_RSP, ctxt->address_size);
+    sp.mem.seg = SEG_SS;
+    sp.value = ctxt->src1.value;
+
+    // Post-increment stack pointer
+    register_add(ctxt, REG_RSP, +(int64_t)ctxt->operand_size);
```

Comment:
The spec requires us to increment after copy. But IIUC, it doesn't really matter if we call `register_add()` before or after `operand_read()`, because the actual register is not updated until `gpr_cache_flush()`. That's cool!

Reply:
Yeah, after setting the appropriate `sp->mem.ea`, we can change anything we want.